bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/CreateAccessPoliciesCommand.cs

```diff
@@ -32,21 +35,33 @@ public class CreateAccessPoliciesCommand : ICreateAccessPoliciesCommand
             throw new NotFoundException();
         }
 
-        var orgAdmin = await _currentContext.OrganizationAdmin(project.OrganizationId);
-        var accessClient = AccessClientHelper.ToAccessClient(_currentContext.ClientType, orgAdmin);
+        await CheckPermissionAsync(project.OrganizationId, userId, projectId);
+        CheckForDistinctAccessPolicies(accessPolicies);
+        await CheckAccessPoliciesDoNotExistAsync(accessPolicies);
 
-        var hasAccess = accessClient switch
-        {
-            AccessClientType.NoAccessCheck => true,
-            AccessClientType.User => await _projectRepository.UserHasWriteAccessToProject(project.Id, userId),
-            _ => false,
-        };
+        await _accessPolicyRepository.CreateManyAsync(accessPolicies);
+        return await _accessPolicyRepository.GetManyByGrantedProjectIdAsync(projectId);
+    }
 
-        if (!hasAccess)
+    public async Task<IEnumerable<BaseAccessPolicy>> CreateForServiceAccountAsync(Guid serviceAccountId,
```

Comment:
Thoughts on combining these two methods? We could fetch the project and service account ids from the access policies and check the access there.

Do we have any logic for checking if someone posts access checks related to a different entity within these requests?

Reply:
It's a creation, so in cases where there are no pre-existing access policies, there is nothing to fetch.

Reply:
"Thoughts on combining these two methods? We could fetch the project and service account ids from the access policies and check the access there."

@Hinton Not a fan of combining the two. I think how it is now is easier to digest.

We would have to have a large switch statement to parse out the granted ID of the target project or service account.

Would look something like this? 

```
using Bit.Core.Context;
using Bit.Core.Enums;
using Bit.Core.Exceptions;
using Bit.Core.SecretsManager.Commands.AccessPolicies.Interfaces;
using Bit.Core.SecretsManager.Entities;
using Bit.Core.SecretsManager.Repositories;

namespace Bit.Commercial.Core.SecretsManager.Commands.AccessPolicies;

public class CreateAccessPoliciesCommand : ICreateAccessPoliciesCommand
{
    private readonly IAccessPolicyRepository _accessPolicyRepository;
    private readonly ICurrentContext _currentContext;
    private readonly IProjectRepository _projectRepository;
    private readonly IServiceAccountRepository _serviceAccountRepository;

    public CreateAccessPoliciesCommand(
        IAccessPolicyRepository accessPolicyRepository,
        ICurrentContext currentContext,
        IProjectRepository projectRepository,
        IServiceAccountRepository serviceAccountRepository)
    {
        _accessPolicyRepository = accessPolicyRepository;
        _currentContext = currentContext;
        _projectRepository = projectRepository;
        _serviceAccountRepository = serviceAccountRepository;
    }
    
    public async Task<IEnumerable<BaseAccessPolicy>> CreateAsync(List<BaseAccessPolicy> accessPolicies, Guid userId)
    {
        var grantedTarget = await GetGrantedTarget(accessPolicies, userId);
        CheckForDistinctAccessPolicies(accessPolicies);
        await CheckAccessPoliciesDoNotExistAsync(accessPolicies);

        await _accessPolicyRepository.CreateManyAsync(accessPolicies);
        if (grantedTarget.Type == "project")
        {
            return await _accessPolicyRepository.GetManyByGrantedProjectIdAsync(grantedTarget.Id);
        }
        return await _accessPolicyRepository.GetManyByGrantedServiceAccountIdAsync(grantedTarget.Id);
    }

    private async Task<GrantedTarget> GetGrantedTarget(List<BaseAccessPolicy> baseAccessPolicies, Guid userId)
    {
        GrantedTarget grantedTarget = null;

        foreach (var baseAccessPolicy in baseAccessPolicies)
        {
            switch (baseAccessPolicy)
            {
                case UserProjectAccessPolicy ap:
                {
                    if (grantedTarget == null && ap.GrantedProjectId != null)
                    {
                        grantedTarget = new GrantedTarget { Id = ap.GrantedProjectId.Value, Type = "project" };
                    }
                    else if (grantedTarget!.Id != ap.GrantedProjectId!.Value)
                    {
                        throw new BadRequestException("Must have the same granted id");
                    }

                    break;
                }
                case GroupProjectAccessPolicy ap:
                {
                    if (grantedTarget == null && ap.GrantedProjectId != null)
                    {
                        grantedTarget = new GrantedTarget { Id = ap.GrantedProjectId.Value, Type = "project" };
                    }
                    else if (grantedTarget!.Id != ap.GrantedProjectId!.Value)
                    {
                        throw new BadRequestException("Must have the same granted id");
                    }

                    break;
                }
                case ServiceAccountProjectAccessPolicy ap:
                {
                    if (grantedTarget == null && ap.GrantedProjectId != null)
                    {
                        grantedTarget = new GrantedTarget { Id = ap.GrantedProjectId.Value, Type = "project" };
                    }
                    else if (grantedTarget!.Id != ap.GrantedProjectId!.Value)
                    {
                        throw new BadRequestException("Must have the same granted id");
                    }

                    break;
                }
                case UserServiceAccountAccessPolicy ap:
                {
                    if (grantedTarget == null && ap.GrantedServiceAccountId != null)
                    {
                        grantedTarget = new GrantedTarget { Id = ap.GrantedServiceAccountId.Value, Type = "serviceAccount" };
                    }
                    else if (grantedTarget!.Id != ap.GrantedServiceAccountId!.Value)
                    {
                        throw new BadRequestException("Must have the same granted id");
                    }

                    break;
                }
                case GroupServiceAccountAccessPolicy ap:
                {
                    if (grantedTarget == null && ap.GrantedServiceAccountId != null)
                    {
                        grantedTarget = new GrantedTarget { Id = ap.GrantedServiceAccountId.Value, Type = "serviceAccount" };
                    }
                    else if (grantedTarget!.Id != ap.GrantedServiceAccountId!.Value)
                    {
                        throw new BadRequestException("Must have the same granted id");
                    }

                    break;
                }
            }
        }

        if (grantedTarget.Type == "project")
        {
            var project = await _projectRepository.GetByIdAsync(grantedTarget.Id);
            if (project == null || !_currentContext.AccessSecretsManager(project.OrganizationId))
            {
                throw new NotFoundException();
            }
            await CheckPermissionAsync(project.OrganizationId, userId, project.Id);
        }
        else
        {
            var serviceAccount = await _serviceAccountRepository.GetByIdAsync(grantedTarget.Id);
            if (serviceAccount == null || !_currentContext.AccessSecretsManager(serviceAccount.OrganizationId))
            {
                throw new NotFoundException();
            }

            await CheckPermissionAsync(serviceAccount.OrganizationId, userId, serviceAccountIdToCheck: serviceAccount.Id);
        }

        return grantedTarget;
    }

    private class GrantedTarget
    {
        public Guid Id { get; set; }
        public string Type { get; set; }
        public Guid OrganizationId { get; set; }
    }

 

    private static void CheckForDistinctAccessPolicies(IReadOnlyCollection<BaseAccessPolicy> accessPolicies)
    {
        var distinctAccessPolicies = accessPolicies.DistinctBy(baseAccessPolicy =>
        {
            return baseAccessPolicy switch
            {
                UserProjectAccessPolicy ap => new Tuple<Guid?, Guid?>(ap.OrganizationUserId, ap.GrantedProjectId),
                GroupProjectAccessPolicy ap => new Tuple<Guid?, Guid?>(ap.GroupId, ap.GrantedProjectId),
                ServiceAccountProjectAccessPolicy ap => new Tuple<Guid?, Guid?>(ap.ServiceAccountId,
                    ap.GrantedProjectId),
                UserServiceAccountAccessPolicy ap => new Tuple<Guid?, Guid?>(ap.OrganizationUserId,
                    ap.GrantedServiceAccountId),
                GroupServiceAccountAccessPolicy ap => new Tuple<Guid?, Guid?>(ap.GroupId, ap.GrantedServiceAccountId),
                _ => throw new ArgumentException("Unsupported access policy type provided.", nameof(baseAccessPolicy)),
            };
        }).ToList();

        if (accessPolicies.Count != distinctAccessPolicies.Count)
        {
            throw new BadRequestException("Resources must be unique");
        }
    }

    private async Task CheckAccessPoliciesDoNotExistAsync(List<BaseAccessPolicy> accessPolicies)
    {
        foreach (var accessPolicy in accessPolicies)
        {
            if (await _accessPolicyRepository.AccessPolicyExists(accessPolicy))
            {
                throw new BadRequestException("Resource already exists");
            }
        }
    }

    private async Task CheckPermissionAsync(Guid organizationId,
        Guid userId,
        Guid? projectIdToCheck = null,
        Guid? serviceAccountIdToCheck = null)
    {
        var orgAdmin = await _currentContext.OrganizationAdmin(organizationId);
        var accessClient = AccessClientHelper.ToAccessClient(_currentContext.ClientType, orgAdmin);

        bool hasAccess;
        switch (accessClient)
        {
            case AccessClientType.NoAccessCheck:
                hasAccess = true;
                break;
            case AccessClientType.User:
                if (projectIdToCheck != null)
                {
                    hasAccess = await _projectRepository.UserHasWriteAccessToProject(projectIdToCheck.Value, userId);
                }
                else if (serviceAccountIdToCheck != null)
                {
                    hasAccess = await _serviceAccountRepository.UserHasWriteAccessToServiceAccount(
                        serviceAccountIdToCheck.Value,
                        userId);
                }
                else
                {
                    hasAccess = false;
                }
                break;
            default:
                hasAccess = false;
                break;
        }

        if (!hasAccess)
        {
            throw new NotFoundException();
        }
    }
}

```

Reply:
As long as we prevent the users from specifying their own we're probably good.

The way I was thinking was to essentially have something like

```c#
CreateAccessPolicies(List<BaseAccessPolicy> accessPolicies, Guid userId) {
  CheckForDistinctAccessPolicies(accessPolicies);

  var projects = accessPolicies.Where(..);
  var serviceAccounts = accessPolicies.Where(..);

  // Check access to projects and service accounts
  // Not duplicates

  await _accessPolicyRepository.CreateManyAsync(accessPolicies);
}